File: src/main/java/com/hp/jipp/encoding/BooleanType.java
package com.hp.jipp.encoding;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/** A boolean attribute type */
public class BooleanType extends AttributeType<Boolean> {
    private static final String TYPE_NAME = "Boolean";

    static final Attribute.SimpleEncoder<Boolean> ENCODER = new Attribute.SimpleEncoder<Boolean>(TYPE_NAME) {
        @Override
        public void writeValue(DataOutputStream out, Boolean value) throws IOException {
            out.writeShort(1);
            out.writeByte(value ? 0x01 : 0x00);
        }

        @Override
        public Boolean readValue(DataInputStream in, Tag valueTag) throws IOException {
            Attribute.expectLength(in, 1);
            return in.readByte() != 0;
        }

        @Override
        public boolean valid(Tag valueTag) {
            return valueTag == Tag.BooleanValue;
        }
    };

    public BooleanType(Tag tag, String name) {
        super(ENCODER, tag, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/BooleanType.kt
package com.hp.jipp.encoding

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** A boolean attribute type */
class BooleanType(tag: Tag, name: String) : AttributeType<Boolean>(BooleanType.ENCODER, tag, name) {
    companion object : IppEncodings {
        private val TYPE_NAME = "Boolean"

        @JvmField
        val ENCODER: Attribute.SimpleEncoder<Boolean> = object : Attribute.SimpleEncoder<Boolean>(TYPE_NAME) {
            @Throws(IOException::class)
            override fun writeValue(out: DataOutputStream, value: Boolean) {
                out.writeShort(1)
                out.writeByte(if (value) 0x01 else 0x00)
            }

            @Throws(IOException::class)
            override fun readValue(input: DataInputStream, valueTag: Tag): Boolean {
                input.takeLength(1)
                return input.readByte().toInt() != 0
            }

            override fun valid(valueTag: Tag): Boolean {
                return valueTag === Tag.BooleanValue
            }
        }
    }
}
